### PR TITLE
CAT Part 4: plumb cat pool to the rest of tendermint

### PR DESCRIPTION
**Pull request by cmwaters**
Ref: #935

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -735,7 +734,7 @@ type MempoolConfig struct {
 // DefaultMempoolConfig returns a default configuration for the Tendermint mempool
 func DefaultMempoolConfig() *MempoolConfig {
 	return &MempoolConfig{
-		Version:   MempoolV0,
+		Version:   MempoolV2,
```
Comment (SweeXordious) [inline config/config.go:737]: [Non Blocking][Question]
Will we be using CAT as soon as its merged by default? shouldn't it be tested for some time first?

File: config/toml.go [hunk 2]
```diff
@@ -345,6 +345,7 @@ dial_timeout = "{{ .P2P.DialTimeout }}"
 # Mempool version to use:
 #   1) "v0" - (default) FIFO mempool.
 #   2) "v1" - prioritized mempool.
+#   3) "v2" - content addressable transaction pool
```
Comment (SweeXordious) [inline config/toml.go:348]: similar 
```suggestion
#   3) "v2" - (default) content addressable transaction pool
```

and  `1) "v0" - (default) FIFO mempool.` =>  `1) "v0" - FIFO mempool.`

**Review by SweeXordious — COMMENTED** [review]
Questions to understand more :D

File: config/config.go (hunk 1 above)
Comment (cmwaters) [reply config/config.go:737]: I have been testing it in testground (and this is the reason why this change exists in the first place) but you're totally right. We should have it opt-in first and v1 be the default

File: mempool/cat/spec.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
+# Content Addressable Transaction Pool Specification
+
+- 01.12.2022 | Initial specification (@cmwaters)
+- 09.12.2022 | Add Push/Pull mechanics (@cmwaters)
+
+### Outline
+
+This document specifies the properties, design and implementation of a content addressable transaction pool (CAT). This protocol is intended as an alternative to the FIFO and Priority mempools currently built-in to the Tendermint consensus protocol. The term content-addressable here, indicates that each transaction is identified by a smaller, unique tag (in this case a sha256 hash). These tags are broadcast among the transactions as a means of more compactly indicating which peers have which transactions. Tracking what each peer has aims at reduces the amount of duplication. In a network without content tracking, a peer may receive as many duplicate transactions as peers connected to. The tradeoff here therefore is that the transactions are significantly larger than the tag such that the sum of the data saved sending what would be duplicated transactions is larger than the sum of sending each peer a tag.
```
Comment (SweeXordious) [inline mempool/cat/spec.md:8]:
```suggestion
This document specifies the properties, design and implementation of a content addressable transaction pool (CAT). This protocol is intended as an alternative to the FIFO and Priority mempools currently built-in to the Tendermint consensus protocol. The term content-addressable here, indicates that each transaction is identified by a smaller, unique tag (in this case a sha256 hash). These tags are broadcast among the transactions as a means of more compactly indicating which peers have which transactions. Tracking what each peer has aims at reducing the amount of duplication. In a network without content tracking, a peer may receive as many duplicate transactions as peers connected to. The tradeoff here therefore is that the transactions are significantly larger than the tag such that the sum of the data saved sending what would be duplicated transactions is larger than the sum of sending each peer a tag.
```

File: mempool/cat/spec.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+Both messages are sent across a new channel with the ID: `byte(0x31)`. This enables cross compatability as discussed in greater detail below.
+
+> **Note:**
+> The term `SeenTx` is used over the more common `HasTx` because the transaction pool contains sophisticated eviction logic. TTL's, higher priority transactions and reCheckTx may mean that a transaction pool *had* a transaction but does not have it any more. Semantically it's more appropriate to use `SeenTx` to imply not the presence of a transaction but that the node has seen it and dealt with it accordingly.
+
+### Outbound logic
+
+A node in the protocol has two distinct modes: "broadcast" and "request/response". When a node receives a transaction via RPC (or specifically through `CheckTx`), it assumed that it is the only recipient from that client and thus will immediately send that transaction, after validation, to all connected peers. Afterwards, only "request/response" is used to disseminate that transaction to everyone else.
```
Comment (SweeXordious) [inline mempool/cat/spec.md:48]: [Question]
If  a transaction is sent to a validator, and that validator is connected to 50 peers. Then, that transaction will only make it to the 50 connected peers. If we assume a validator set of 150, 100 validators will not see it until it is included in a block. Won't this delay the amount of time users will need to wait before their transaction gets included?

Comment (cmwaters) [reply mempool/cat/spec.md:48]: > If a transaction is sent to a validator, and that validator is connected to 50 peers. Then, that transaction will only make it to the 50 connected peers.

Only initially. Then each of those 50 connected peers will broadcast `SeenTx` messages to the peers they are connected with which means all nodes should eventually see these `SeenTx` messages. For the nodes that don't have the transaction. They can send a `WantTx` message, essentially requesting it from that peer. That way all nodes will eventually receive all transactions.

Comment (SweeXordious) [reply mempool/cat/spec.md:48]: Thanks a lot :+1:

File: config/config.go (hunk 1 above)
Comment (evan-forbes) [reply config/config.go:737]: I'm ok with being degens and leaving this as the default, wcgw? :cat2:

Comment (evan-forbes) [reply config/config.go:737]: jk jk

**Review by evan-forbes — DISMISSED** [review]
none other than what was already discussed in the original specs PR! LGTM

let's :shipit: CAT

**Review by evan-forbes — APPROVED** [review]

**Pull request merged**